### PR TITLE
refactor: Add unified JSON interface

### LINKED ISSUE #3264: Make JSON values use an interface
In our code base we use normalized values to represent values of doc fields. JSON can also be a value, but it's content is not normalized at the moment, e.g. JSON number can be `int`, `float64` or other variations.

We should make all JSON values implement a common interface similar to `NormalValue` interface.

This will also allow us to do more sophisticated operation with JSON, for example, for indexing of json values.

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #3264

## Description

Introduces a common interface for JSON values that would allow later to do better processing of json objects.

### CODE REVIEW

**Review by AndrewSisley — COMMENTED** [review]
Looks good to me - just a couple of small comments from me :)

question: What was the motivation for doing this now?  Is it going to help the development of a feature you are working on, or did you just feel in the mood?

File: client/json.go
Comment (fredcarle) [reply client/json.go:260]: I disagree. `ParseBytes` calls `Parse` internally so it would be weird to go from string to only to go back to string.

Comment (AndrewSisley) [reply client/json.go:260]: Ah, well we can to the inverse then no?
```
func ParseJSONBytes(data string) (JSON, error) {
  return ParseJSONString([]byte(data)
}
```
The important part being that they are exactly the same/interchangable.

Comment (AndrewSisley) [reply client/json.go:260]: Or if you'd argue that has the same problem and that it is the cast that you dislike - I disagree, as the cast is basically free and not much more than a compiler type-hint than anything serious at runtime, and in exchange we get a very clear guarantee and documentation that the two functions have the same behaviour.

Comment (fredcarle) [reply client/json.go:260]: The thing is that it's more efficient to call the `ParseBytes` for bytes instead of casting bytes as string. `fastjson` uses an unsafe pointer to avoid copying memory which can be a lot more efficient for large JSON objects. If you absolutely want them to be the same we should do like they do.
```go
func PardeJSONBytes(data []byte) (JSON, error) {
    return ParseJSONString(*(*string)(unsafe.Pointer(&data)))
}
```

Comment (AndrewSisley) [reply client/json.go:257]: Casting between `string` and `[]byte` copies in Go?  I'm happy as-is then :)

Comment (islamaliev) [reply client/json.go:257]: > Casting between string and []byte copies in Go

yes, in both directions

Comment (AndrewSisley) [reply client/json.go:257]: > That's what I meant by "fastjson.Parser.ParseBytes casts the bytes slice to a string internally"

Sorry - I got your comment notification after Fred's and his reply - I didn't mean to ignore you :)

Comment (AndrewSisley) [reply client/json.go:257]: > yes, in both directions

Thanks!

File: client/value.go
Comment (AndrewSisley) [reply client/value.go:36]: Sorry for the late suggestion, I could have (but didn't) spot this earlier:

suggestion: Would it not be more correct to host this logic in `NormalValue`, and change this func to:
```
func (val FieldValue) Value() any {
  return val.value.Unwrap()
}
```
Otherwise `NormalValue`.`Unwrap` is not fully unwrapping the values.

Comment (islamaliev) [reply client/value.go]: turns out json normal value is already doing this, so all this stuff in `FieldValue` was not necessary.
It seems like I put them there because I was confused a bit by errors due to different `int64` vs `float64` types and thought that would fix it.
Anyway thanks a lot, Andy, for helping clarify this stuff.

File: client/json.go
Comment (fredcarle) [inline client/json.go:58]: question: Why did you want to support mashalling to a writer?

File: client/value.go
Comment (AndrewSisley) [reply client/value.go]: Ah sweet, thanks Islam :)

**Review by AndrewSisley — APPROVED** [review]
LGTM, but please wait for Fred to finish reviewing before merge :)

File: client/json.go
Comment (fredcarle) [inline client/json.go:369]: question: You prefer
```go
arr := make([]foo, 0, len(bar))
for _, item := range bar {
    arr = append(arr, newFoo(item))
}
```
to?
```go
arr := make([]foo, len(bar))
for i, item := range bar {
    arr[i] = newFoo(item)
}
```

Not pushing back. Just wondering.

**Review by fredcarle — COMMENTED** [review]
LGTM.  Looking forward to see how this will be used for indexing JSON :)

**Review by fredcarle — APPROVED** [review]
Did comment instead of approve 🙃

File: client/json.go
Comment (islamaliev) [reply client/json.go:58]: I though that if someone wants to do some fancy marshaling with a complex json object he would pass around a buffer instead of calling `MarshalJSON` and appending results all the time.

**Pull request merged**
